src/pool/srv_pool_scrub.c

```diff
@@ -4,324 +4,771 @@
+	 * loop through each checksum and chunk of the recx based
+	 * on chunk size.
+	 */
+	for (i = 0; i < csum_nr; i++) {
+		uint8_t		*orig_csum = NULL;
+		bool		 match;
+		daos_size_t	 rec_in_chunk;
+
+		orig_csum = ci_idx2csum(ctx->sc_csum_to_verify, i);
+		rec_in_chunk = sc_get_rec_in_chunk_at_idx(ctx, i);
+
+		/** set an iov with just the data for the current chunk */
+		d_iov_set(&chunk_iov, data->iov_buf + processed_bytes,
+			  rec_in_chunk * rec_len);
+		D_ASSERT(processed_bytes + chunk_iov.iov_len <= data->iov_len);
+
+		rc = daos_csummer_calc_for_iov(sc_csummer(ctx), &chunk_iov,
+					       csum_buf, csum_len);
+		if (rc != 0) {
+			D_ERROR("daos_csummer_calc_for_iov error: "DF_RC"\n",
+				DP_RC(rc));
+			D_GOTO(done, rc);
+		}
+
+		match = daos_csummer_csum_compare(sc_csummer(ctx), orig_csum,
+						  csum_buf, csum_len);
+
+		if (!match) {
+			D_ERROR("Corruption found for chunk #%d of recx: "
+					DF_RECX"\n", i, DP_RECX(*recx));
+			D_GOTO(done, rc = -DER_CSUM);
+		}
+
+		processed_bytes += chunk_iov.iov_len;
+
+		sc_verify_finish(ctx);
+	}
+
+done:
+	D_FREE(csum_buf);
+
+	return rc;
+}
+
+static int
+sc_verify_sv(struct scrub_ctx *ctx, d_sg_list_t *sgl)
+{
+	int rc;
+
+	rc = daos_csummer_verify_key(sc_csummer(ctx), &sgl->sg_iovs[0],
+				     ctx->sc_csum_to_verify);
+	sc_verify_finish(ctx);
+
+	return rc;
+}
+
+static uuid_t *
+sc_cont_uuid(struct scrub_ctx *ctx)
 {
-	char *enabled = getenv("DAOS_CSUM_SCRUB");
+	return &ctx->sc_cont.scs_cont_uuid;
+}
+
+static daos_handle_t
+sc_cont_hdl(struct scrub_ctx *ctx)
+{
+	return ctx->sc_cont.scs_cont_hdl;
+}
 
-	return enabled != NULL && strncmp(enabled, "ON", strlen("ON")) == 0;
+static void
+sc_raise_ras(struct scrub_ctx *ctx)
+{
+	if (ds_notify_ras_event != NULL) {
+		ds_notify_ras_event(RAS_CORRUPTION_DETECTED,
+				    "Data corruption detected",
+				    RAS_TYPE_INFO,
+				    RAS_SEV_ERROR, NULL, NULL, NULL, NULL,
+				    &ctx->sc_pool_uuid, sc_cont_uuid(ctx),
+				    NULL, NULL, NULL);
+	}
+}
+
+static int
+sc_mark_corrupt(struct scrub_ctx *ctx)
+{
+	return vos_iter_process(ctx->sc_vos_iter_handle,
+				VOS_ITER_PROC_OP_MARK_CORRUPT, NULL);
 }
 
-static uint64_t
-interval_sec()
+static int
+sc_verify_obj_value(struct scrub_ctx *ctx)
 {
-	char *sec = getenv("DAOS_CSUM_SCRUB_INTERVAL_SEC");
+	d_sg_list_t	 sgl;
+	daos_iod_t	*iod = &ctx->sc_iod;
+	uint64_t	 data_len;
+	int		 rc;
+
+	D_DEBUG(DB_CSUM, "Scrubbing iod: "DF_C_IOD"\n", DP_C_IOD(iod));
+	/*
+	 * Know that there will always only be 1 recx because verifying a
+	 * single extent at a time so use first recx in iod for data_len
+	 * calculation
+	 */
+	data_len = iod->iod_type == DAOS_IOD_ARRAY ?
+		   iod->iod_recxs[0].rx_nr * iod->iod_size :
+		   iod->iod_size;
+	/* allocate memory to fetch data into */
+	d_sgl_init(&sgl, 1);
+	D_ALLOC(sgl.sg_iovs[0].iov_buf, data_len);
+	sgl.sg_iovs[0].iov_buf_len = data_len;
+	sgl.sg_iovs[0].iov_len = data_len;
+
+	/* Fetch data */
+	rc = vos_obj_fetch(sc_cont_hdl(ctx), ctx->sc_cur_oid,
```

Comment:
I like this idea. I think it will mean that I need to move srv_pool_scrub to the vos layer so it can get to the bio_context.
I've really tried to keep server dependencies out of it so that might be okay. I'll look into it.

Reply:
Updated so srv_pool_scrub is now vos_pool_scrub and changed from vos_obj_read to bio_read()